### PR TITLE
MB-36: persist batch/expiry data on kreditor order line save, block r…

### PR DESCRIPTION
**Pull request by ZaynSaul**
## Summary
Kreditor goods receipt with batch/expiry tracking was completely broken: when a user filled
in Udløbsdato (expiry date) and Batchnr. on a purchase order line and saved, those two
values were silently discarded. The columns existed on `ordrelinjer`, they were read back
everywhere downstream (display, goods receipt, `batch_kob`), but nothing ever wrote the
submitted `$_POST` values into `ordrelinjer` on save — a save-step was simply missing.

Result: `lager/batch_oversigt.php` always showed blank Batchnr./Udløbsdato, and receipt went
through with no warning that batch data was missing.

### 1. The core bug — kreditor/ordre.php
Added the missing save step: captured `$_POST['batch_due_date']`/`'batch_batch_no'` and
included them in the `UPDATE ordrelinjer` that saves a kladde line.

### 2. The ticket's secondary ask — kreditor/modtag.php
Added a hard stop mirroring the existing "Serienumre ikke udfyldt" check: receiving a
batch/expiry-tracked line with an empty due-date or batch-no now blocks with "Batchoplysninger
ikke udfyldt" instead of silently receiving it blank.

### 3. Out-of-scope bug found and fixed with explicit approval — includes/stdFunc/fefo.php
While live-testing the fixes in (1)/(2) above, discovered the batch/expiry input fields never rendered at
all for any item, regardless of the "Varen har udløbsdato" checkbox on the item card.

**Root cause:** `item_has_due_date()` was checking `grupper.box9` (an unrelated, group-level
"batch control" flag used by a different subsystem) instead of `varer.has_due_date` (the
actual per-item checkbox). Git history showed this was a deliberate regression introduced by
an unrelated commit ("batch management setup completed") that overwrote a previously-correct
implementation.

**This was flagged explicitly before being touched**, since it's a separate bug outside
MB-36's stated scope. Approved for inclusion in this same PR, since without it the whole
feature (the actual subject of MB-36) is unreachable — fixing (1)/(2) above alone would have no
visible effect. Restored the function to check `varer.has_due_date`.

### 4. Follow-up fixes from a code review pass (same three files, no new files)
- `ordre.php`: guarded the new POST reads with `ifset()` — raw array access threw new PHP 8
  "undefined array key"/"access on null" warnings on any save where a line's item isn't
  batch-tracked (the common case).
- `modtag.php` (4 call sites) + `ordre.php`: fixed `batch_due_date` being interpolated into
  SQL unescaped while the sibling `batch_batch_no` was escaped — inconsistent, and this fix
  is what first makes this field carry live data, so the gap was latent until now.
- Both files: fixed a `!$value` truthy check that treated a legitimate batch number of `"0"`
  as empty, wrongly blocking receipt or silently storing `NULL`.

## Files Changed
- kreditor/ordre.php — save-side persistence + null/empty-safety fixes
- kreditor/modtag.php — receipt validation + escaping/empty-string fixes (4 sites)
- includes/stdFunc/fefo.php — restored `item_has_due_date()` to the correct per-item check
  (approved out-of-scope fix, see #3 above)

## How to Test
1. On an item with "Varen har udløbsdato" checked on the item card, open a kreditor purchase
   order line for that item — verify the Udløbsdato/Batchnr. input fields now render (they
   previously never did, regardless of this checkbox).
2. Fill in Udløbsdato and Batchnr. on the line, save — verify both values persist on reload
   (previously silently discarded).
3. Approve and receive the order — verify `batch_kob` is populated correctly with the saved
   due date and batch number.
4. Check `lager/batch_oversigt.php` — verify it now correctly displays the batch data.
5. Attempt to receive a batch/expiry-tracked line with Udløbsdato or Batchnr. left empty —
   verify it's blocked with "Batchoplysninger ikke udfyldt," matching the existing
   "Serienumre ikke udfyldt" pattern.
6. Test the `"0"` batch-number edge case specifically — enter a batch number of literally
   `"0"` and verify it saves and receives correctly, not treated as empty/blocked.
7. Save a kladde line for an item that is NOT batch/expiry-tracked — verify no PHP warnings
   are emitted (regression check for the `ifset()` guard fix).
8. Check the Apache error log after a full save → approve → receive cycle — confirm no new
   PHP warnings anywhere in the flow.
9. Run `php -l` on all three changed files — confirm clean.
10. Run `composer test` — confirm no change from the pre-existing baseline.

## Acceptance Criteria
| # | Scenario | Expected Result |
|---|---|---|
| 1 | Item with due-date tracking enabled, kreditor order line | Batch/expiry input fields render correctly |
| 2 | Fill Udløbsdato/Batchnr. on a line and save | Values persist, not discarded |
| 3 | Approve and receive the order | batch_kob populated correctly with saved values |
| 4 | lager/batch_oversigt.php | Displays batch data correctly |
| 5 | Receive a batch/expiry-tracked line with missing due-date/batch-no | Blocked with "Batchoplysninger ikke udfyldt" |
| 6 | Batch number of "0" | Saved and received correctly, not treated as empty |
| 7 | Save a non-batch-tracked item's line | No PHP warnings |
| 8 | Full save → approve → receive cycle | No new PHP warnings in Apache error log |

## Verification
- `php -l` clean on all three files.
- `composer test` unchanged from the pre-existing baseline (confirmed via `git stash`
  comparison).
- Full end-to-end run performed twice against the real app in the `chartest` tenant (once
  via direct tenant login, once via the master admin's "View accounts" panel): save →
  approve → receive → `batch_kob` populated correctly → `lager/batch_oversigt.php` displays
  the batch data — including the `"0"` batch-number edge case.
- Confirmed via the Apache error log that no new PHP warnings are emitted.

## Have you checked the following? 
- [ ] I have updated relevant documentation at https://github.com/DANOSOFT/saldi/wiki
- [x] I have checked that i am not linking to any external resources such as external style- or JavaScript-files, that could compromise stability
- [x] I have read and understood the developer guidelines  
      https://docs.google.com/document/d/1GOmomtvKf21OV2VWNOIPweDi4gJHpMCK5qXzrPrypUc/edit?usp=sharing

## Summary by CodeRabbit

- **Bug Fixes**
  - Due-date and batch tracking now reflects each item’s configured settings across inventory workflows.
  - Receiving prevents completion when required batch numbers or due dates are missing.
  - Batch number `0` is handled as a valid value.
  - Open creditor orders now correctly display batch fields for tracked items.

- **Improvements**
  - Batch due dates and batch numbers entered on creditor orders are validated and saved with order lines.

### CODE REVIEW

File: includes/stdFunc/fefo.php [hunk 1]
```diff
@@ -65,17 +65,16 @@ function fefo_batch_query($vare_id, $lager = 0, $variant_id = 0) {
...
-	 * @return bool  True if the group has batch tracking enabled
+	 * @return bool  True if the item has due_date tracking enabled
 	 */
 	function item_has_due_date($vare_id) {
-		global $regnaar;
-		$qtxt = "SELECT g.box9 FROM varer v JOIN grupper g ON g.kodenr = v.gruppe AND g.art = 'VG' AND g.fiscal_year = '$regnaar' WHERE v.id = '$vare_id'";
+		$qtxt = "SELECT has_due_date FROM varer WHERE id = '$vare_id'";
 		$r = db_fetch_array(db_select($qtxt, __FILE__ . " linje " . __LINE__));
-		return ($r && trim($r['box9']) === 'on');
+		return ($r && ($r['has_due_date'] === 't' || $r['has_due_date'] === true || $r['has_due_date'] == 1));
```
[the remark this answers is not in the document]
Comment (ZaynSaul) [reply includes/stdFunc/fefo.php:77]: Verified this against the actual code paths before deciding whether to change anything:

Every combination of `box8`/`box9` already ends in a `batch_kob` insert/update for the received line — the direct inserts at `modtag.php:207-211` and `:244-246` for non-`box9` paths, and via `reservation()`/`returnering()` for the `box9=='on'` path. This is pre-existing, universal behavior unrelated to `has_due_date` — every received line gets a `batch_kob` row (it's the general purchase-lot table used for FEFO/cost basis), with `due_date`/`batch_no` simply NULL when the item has no expiry data. That was true before this PR and is unchanged by it.

`box9` and `has_due_date` govern two different things: `box9` selects *how* the batch_kob/lagerstatus bookkeeping happens for a group (direct insert vs. via `reservation()`/`returnering()`), while `has_due_date` is the item-level "does this item actually carry expiry dates" flag that gates the UI fields and my new validation. A `box9=='on'` item with `has_due_date=false` still gets correctly batch-posted (lot-tracked), just with a NULL expiry date — which is the correct state for an item that genuinely has no expiry date. Live-verified: receiving such an item does not error, and its `batch_kob` row carries the right non-null fields with `due_date`/`batch_no` NULL.

So I don't think there's a live defect here — `has_due_date` correctly gates *expiry-date requirements*, and `box9` correctly gates the *stock-posting mechanism*, and nothing requires them to be the same flag. Backfilling/syncing them would conflate two independent concerns without a concrete failure case driving it. Not changing this in MB-36's scope — happy to revisit if you can point to a specific box9 group where items are expected to always carry expiry dates that this would miss.

**Comment by LuiHedlund02** [comment]
Lead reviewer: @SirRolin (already requested)
Reason: changes which items get batch/expiry fields and adds a hard stop on goods receipt.
Focus: `includes/stdFunc/fefo.php` now reads `varer.has_due_date`, a column added only by `opdat_4.2`; the 4.2/4.3 gate was dead on some tenants (see #564), so confirm the column exists on a tenant that never ran that opdat. Also the `grupper.box9` vs `varer.has_due_date` predicate split: a group-tracked item with the per-item flag off renders no fields and is not validated, but still writes `batch_kob`. In-flight orders hit the new receipt stop with no backfill.
Verification: author reports two end-to-end runs on chartest incl. batch "0"; no automated test for the new validation stop.
Next action: review now. Merge order vs #457 (same files) to be settled afterwards.

**Comment by LuiHedlund02** [comment]
@SirRolin retaining you as lead for the purchase-receipt batch/expiry change. Review focus: tenant availability of `varer.has_due_date`, item-card vs group predicates (`has_due_date` / box9), existing-item backfill and whether receipt validation blocks previously valid workflows. The 4.2 update gate is already corrected on master; verify schema availability on an upgraded tenant. Author evidence is in the PR; no new live receipt run was performed in this queue pass. Keep #457 behind this change.

**Comment by ZaynSaul** [comment]
@LuiHedlund02 — addressed all three points from your review focus:

**`varer.has_due_date` tenant availability**: Confirmed #564 (the general opdat-gate fix) is still open/unmerged, so it doesn't cover this. Added a direct backfill in `includes/betweenUpdates.php` for all six FEFO columns this feature depends on (`batch_kob.due_date`/`batch_no`, `varer.has_due_date`/`default_shelf_life_days`, `ordrelinjer.batch_due_date`/`batch_batch_no`) — same idempotent existence-check pattern already used for `pool_files.norm_amount`. Runs regardless of a tenant's exact `tjek4opdat.php` version-stepper state.

**`box9` vs `has_due_date` predicate split**: You were right to keep pushing on this — the earlier CodeRabbit thread reply argued this was fine because a NULL `due_date` on a box9-only item is correct (no expiry needed), which is true but missed your actual point: the *batch number* field itself never renders for those items, so they have no way to ever record a lot number at all, even though `box9` was specifically why they went through the batch-tracked receipt path. Fixed by splitting the two fields' gating: `batch_batch_no` now shows/is required when `box9=='on'` OR `item_has_due_date()`, while `batch_due_date` stays gated on `item_has_due_date()` alone. Live-verified on chartest: created a box9-tracked, non-has_due_date item, confirmed the batch-no field renders, confirmed receipt blocks with it empty and succeeds with it filled (batch_kob row created with `batch_no` set, `due_date` correctly NULL), and regression-checked the original has_due_date flow (including the batch_no `"0"` edge case) still works unchanged.

**In-flight orders hitting the new receipt stop**: Looked into this — the batch input fields already existed on master before this PR; what's new is `modtag.php` actually enforcing them. Any order line left blank before this ships will need one-time data entry at receipt, same as any new required-field rollout. Didn't see a clean way to add a grace-period exemption that wouldn't just reintroduce the gap for a while, so leaving this as expected behavior rather than coding around it — flag if you disagree.

**Merge order vs #457**: Checked - no line-level overlap with #608 on `kreditor/ordre.php`, so no action needed on this side; just needs sequencing/rebase whichever merges second.

Added `tests/characterization/kreditor/ModtagBatchValidationTest.test.php` (12 cases) covering the split validation logic. Ready for another look.

File: kreditor/orderIncludes/openOrderLines.php [hunk 2]
```diff
@@ -176,40 +181,78 @@
...
           $shelf_days = item_default_shelf_life($vare_id[$x]);
           if ($shelf_days) $batch_due_date_val = date('Y-m-d', strtotime("+$shelf_days days"));
         }
-        print "<td title='".findtekst('5001|Udl&oslash;bsdato', $sprog_id)."'>";
-        print "<input class='inputbox' type='date' style='width:130px;' name='batch_due_date[$x]' value='$batch_due_date_val' onchange='javascript:docChange = true;'></td>\n";
-        print "<td title='".findtekst('5005|Batchnr.', $sprog_id)."'>";
-        print "<input class='inputbox' type='text' style='width:90px;' name='batch_batch_no[$x]' value='$batch_batch_no_val' onchange='javascript:docChange = true;'></td>\n";
+        if ($dueTracked) {
+          print "<td title='".findtekst('5001|Udl&oslash;bsdato', $sprog_id)."'>";
+          print "<input class='inputbox' type='date' style='width:130px;' name='batch_due_date[$x]' value='$batch_due_date_val' onchange='javascript:docChange = true;'></td>\n";
```
[the remark this answers is not in the document]
Comment (ZaynSaul) [reply kreditor/orderIncludes/openOrderLines.php:209]: Already fixed in 832d246b — both batch_batch_no print sites now use an HTML-escaped variable (`$batch_batch_no_attr = htmlspecialchars((string) $batch_batch_no_val, ENT_QUOTES, 'UTF-8')`), see current lines 205 and 249. This comment's diff hunk is from before that fix landed; the line-number mapping just didn't get marked outdated.

[the remark this answers is not in the document]
Comment (ZaynSaul) [reply kreditor/orderIncludes/openOrderLines.php:209]: Correction on line numbers: the escaped variable is assigned at lines 206/250 and used in the print statements at lines 215/259.

File: includes/betweenUpdates.php [hunk 3]
```diff
@@ -653,6 +656,40 @@
...
+// step (the common case for any live install) never runs it, so the FEFO/batch-expiry columns
+// it adds (batch_kob.due_date/batch_no, varer.has_due_date/default_shelf_life_days,
+// ordrelinjer.batch_due_date/batch_batch_no) can be permanently missing even though the
+// tenant's version implies the feature should be available. Same dead-gate class of bug as the
+// pool_files.norm_amount fix above; backfilled here unconditionally so the feature degrades to
+// "off" instead of silently failing partway through (e.g. ordre.php's save UPDATE referencing a
+// nonexistent ordrelinjer column).
+$qtxt = "SELECT column_name FROM information_schema.columns WHERE table_name='batch_kob' AND column_name='due_date'";
+if (!db_fetch_array(db_select($qtxt, __FILE__ . " linje " . __LINE__))) {
+	db_modify("ALTER TABLE batch_kob ADD COLUMN due_date DATE NULL", __FILE__ . " linje " . __LINE__);
```
[the remark this answers is not in the document]
Comment (ZaynSaul) [reply includes/betweenUpdates.php]: Fixed - the FEFO column backfill now uses the same GET_LOCK/ADD COLUMN IF NOT EXISTS pattern as performed_by and the pool_files vendor columns just above it in this file: one lock covers all six missing columns across batch_kob/varer/ordrelinjer, rechecked under the lock before altering.

File: includes/stdFunc/fefo.php [hunk 4]
```diff
@@ -65,17 +65,16 @@ function fefo_batch_query($vare_id, $lager = 0, $variant_id = 0) {
...
+	 * (varer.has_due_date, set via the "Varen har udl&oslash;bsdato" checkbox on the item card).
 	 *
 	 * @param int $vare_id  Item ID
-	 * @return bool  True if the group has batch tracking enabled
+	 * @return bool  True if the item has due_date tracking enabled
 	 */
 	function item_has_due_date($vare_id) {
-		global $regnaar;
-		$qtxt = "SELECT g.box9 FROM varer v JOIN grupper g ON g.kodenr = v.gruppe AND g.art = 'VG' AND g.fiscal_year = '$regnaar' WHERE v.id = '$vare_id'";
+		$qtxt = "SELECT has_due_date FROM varer WHERE id = '$vare_id'";
```
[the remark this answers is not in the document]
Comment (ZaynSaul) [reply includes/stdFunc/fefo.php]: Fixed - item_has_due_date() now rejects anything that isn't a valid int before it reaches the query.

File: kreditor/orderIncludes/openOrderLines.php [hunk 5]
```diff
@@ -176,40 +184,82 @@
...
+      } else {
+        print "<td></td>";
+      }
+      if ($batchTracked) {
+        print "<td title='".findtekst('5005|Batchnr.', $sprog_id)."'>";
+        print "<input class='inputbox' type='text' style='width:90px;' name='batch_batch_no[$x]' value='$batch_batch_no_attr' onchange='javascript:docChange = true;'></td>\n";
+      } else {
+        print "<td></td>";
+      }
     }
```
[the remark this answers is not in the document]
Comment (ZaynSaul) [reply kreditor/orderIncludes/openOrderLines.php]: Fixed - added the OR v.has_due_date = TRUE condition, so a due-date-tracked item under a box9-off group no longer skips the batch input block.

File: kreditor/ordre.php [hunk 6]
```diff
@@ -868,10 +873,14 @@ function fejltekst(tekst) {
...
-					$qtxt = "update ordrelinjer set beskrivelse='$beskrivelse[$x]', antal='$antal[$x]', leveres='$leveres[$x]', ";
+						$_batch_due_date_raw = ifset($batch_due_date, $x);
+						$_batch_batch_no_raw = ifset($batch_batch_no, $x);
+						$_batch_due_date = ($_batch_due_date_raw !== null && $_batch_due_date_raw !== '') ? "'" . db_escape_string($_batch_due_date_raw) . "'" : 'NULL';
+						$_batch_batch_no = ($_batch_batch_no_raw !== null && $_batch_batch_no_raw !== '') ? "'" . db_escape_string($_batch_batch_no_raw) . "'" : 'NULL';
+						if ($rabat[$x] === '' || $rabat[$x] === null) $rabat[$x] = 0;
+						$qtxt = "update ordrelinjer set beskrivelse='$beskrivelse[$x]', antal='$antal[$x]', leveres='$leveres[$x]', ";
 						$qtxt.= "leveret='$tidl_lev[$x]', pris='$pris[$x]', rabat='$rabat[$x]', projekt='$projekt[$x]',  ";
-						$qtxt.= "omvbet='$omvbet[$x]',lager='$lager' where id='$linje_id[$x]'";
+						$qtxt.= "omvbet='$omvbet[$x]',lager='$lager', batch_due_date=$_batch_due_date, batch_batch_no=$_batch_batch_no where id='$linje_id[$x]'";
```
[the remark this answers is not in the document]
Comment (ZaynSaul) [reply kreditor/ordre.php]: Fixed - batch_due_date/batch_batch_no are now only included in the UPDATE when that line's input was actually submitted, so a line whose tracking flag turned off no longer gets its stored batch data overwritten with NULL.

**Comment by ZaynSaul** [comment]
@coderabbitai review
